File: docs-examples/example-java/src/test/java/io/micronaut/rabbitmq/docs/exchange/CustomExchangeSpec.java
package io.micronaut.rabbitmq.docs.exchange;

import io.micronaut.context.annotation.Property;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;
import io.micronaut.rabbitmq.testcontainers.RabbitMQ;

@MicronautTest
@Property(name = "spec.name", value = "CustomExchangeSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class CustomExchangeSpec implements TestPropertyProvider {
    @Override
    public Map<String, String> getProperties() {
        return RabbitMQ.getProperties();
    }

    @Test
    void testUsingACustomExchange(AnimalClient client, AnimalListener listener) {

        client.send(new Cat("Whiskers", 9));
        client.send(new Cat("Mr. Bigglesworth", 8));
        client.send(new Snake("Buttercup", false));
        client.send(new Snake("Monty the Python", true));

        await().atMost(60, SECONDS).until(() ->
                listener.receivedAnimals.size() == 4 &&
                listener.receivedAnimals.stream()
                        .filter(Cat.class::isInstance)
                        .anyMatch(cat -> cat.getName().equals("Whiskers") && ((Cat) cat).getLives() == 9) &&
                listener.receivedAnimals.stream()
                        .filter(Cat.class::isInstance)
                        .anyMatch(cat -> cat.getName().equals("Mr. Bigglesworth") && ((Cat) cat).getLives() == 8) &&
                listener.receivedAnimals.stream()
                        .filter(Snake.class::isInstance)
                        .anyMatch(snake -> snake.getName().equals("Buttercup") && !((Snake) snake).isVenomous()) &&
                listener.receivedAnimals.stream()
                        .filter(Snake.class::isInstance)
                        .anyMatch(snake -> snake.getName().equals("Monty the Python") && ((Snake) snake).isVenomous())
        );
    }
}

File: docs-examples/example-python/src/test/python/micronaut/rabbitmq/docs/exchange/CustomExchangeSpec.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.rabbitmq.docs.Await import Await
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .AnimalClient import AnimalClient
from .AnimalListener import AnimalListener
from .Cat import Cat
from .Snake import Snake


@MicronautTest(environments=["rabbitmq"])
@Property(name="spec.name", value="CustomExchangeSpec")
class CustomExchangeSpec:
    client: Annotated[AnimalClient, Inject]
    listener: Annotated[AnimalListener, Inject]

    @Test
    def test_using_a_custom_exchange(self) -> None:
        self.client.send_animal(Cat("Whiskers", 9))
        self.client.send_animal(Cat("Mr. Bigglesworth", 8))
        self.client.send_animal(Snake("Buttercup", False))
        self.client.send_animal(Snake("Monty the Python", True))

        received = self.listener.received_animals
        cats = lambda: [cat for cat in received if isinstance(cat, Cat)]
        snakes = lambda: [snake for snake in received if isinstance(snake, Snake)]
        Await.until(lambda: len(received) == 4
                    and any(cat.name == "Whiskers" and cat.lives == 9 for cat in cats())
                    and any(cat.name == "Mr. Bigglesworth" and cat.lives == 8 for cat in cats())
                    and any(snake.name == "Buttercup" and not snake.venomous for snake in snakes())
                    and any(snake.name == "Monty the Python" and snake.venomous for snake in snakes()))
